### PR TITLE
Add simple row-group-aggregating parquet reader

### PR DESCRIPTION
**Pull request by rjzamora**
This is an experimental PR to add a simple `opt_parquet.parquet_reader` routine to `dask_cudf`.  This routine is very limited comparted to the upstream read_parquet, but can more-efficiently aggregate multiple row groups into each output partitions.

Usage example (assigning 16 row-groups to each partition):

`dask_cudf.read_parquet(path, row_groups_per_part=16)`

**Notes**:

- The `path` must be a directory containing a `"_metadata" file (for now).
- Row-group aggregation will only occur within the same file. So, you may have some small row groups if the number of row-groups in each file is not divisible by the `row_groups_per_part` argument.

cc @VibhuJawa @beckernick

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by rjzamora** [comment]
Since this approach requires the directory to contain a `"_metadata"` file, here is an example of creating one after the fact (if the dataset already exists):

```python
import pyarrow.parquet as pq
from fsspec.core import get_fs_token_paths
def write_metadata_file(path, out_path, fs):
    """ Write "_metadata" file for simple parquet dataset
        Notes: Requires a directory path as input.
        Only works for a "flat" directory structure.
    """
    dataset = pq.ParquetDataset(path, filesystem=fs)
    pieces = dataset.pieces
    meta = None
    for piece in pieces:
        _meta = piece.get_metadata()
        filename = piece.path.split(fs.sep)[-1]
        _meta.set_file_path(filename)
        if meta:
            meta.append_row_groups(_meta)
        else:
            meta = _meta
    if meta:
        metadata_path = fs.sep.join([out_path, "_metadata"])
        with fs.open(metadata_path, "wb") as fil:
            meta.write_metadata_file(fil)
    else:
        print("No metadata to write!")

dir_path = "/.../dataset-directory"
out_dir_path = dir_path
fs = get_fs_token_paths(dir_path, mode="rb")[0]
write_metadata_file(dir_path, out_dir_path, fs)
```

**Comment by rjzamora** [comment]
To further motivate the need for a more-efficient row-group aggregation:

```python
import dask_cudf
path = "/raid/dask_space/criteo/crit_pq_int"

# Default
%time dask_cudf.read_parquet(path).npartitions

# Using chunksize
%time dask_cudf.read_parquet(path, chunksize="500MB").npartitions

# Using row_groups_per_part
%time dask_cudf.read_parquet(path, row_groups_per_part=10).npartitions
```

Default:
```
CPU times: user 40.9 s, sys: 977 ms, total: 41.9 s
Wall time: 41.2 s
6337
```

Using `chunksize`:
```
CPU times: user 41.3 s, sys: 1.39 s, total: 42.7 s
Wall time: 42.3 s
612
```

Using `row_groups_per_part`:
```
CPU times: user 675 ms, sys: 68 ms, total: 744 ms
Wall time: 742 ms
644
```

**Comment by quasiben** [comment]
Is this ready for review ?  Anything else you want to push on here ?

**Comment by rjzamora** [comment]
> Is this ready for review ? Anything else you want to push on here ?

Just wanted to add testing and do a bit of cleanup before asking for a review.  Sorry - ended up getting sidetracked by other things so far today.

**Comment by quasiben** [comment]
This looks really really great @rjzamora .  I think it's the first use of `ParquetDataset`, no ?  @kkraus14 do you have any comments here ?

**Comment by rjzamora** [comment]
> This looks really really great @rjzamora . I think it's the first use of ParquetDataset, no ? @kkraus14 do you have any comments here ?

This uses `ParquetDataset`, which is different than the new Arrow Dataset API. We should certainly use that API once it supports `"_metadata"` handling.  Note that @jorisvandenbossche 
 just opened [dask#6174](https://github.com/dask/dask/issues/6174) to discuss the integration of the new API into upstream dask.  We should certainly be involved in this effort and facilitate proper integration however possible :)

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import math
```
Comment (kkraus14) [inline python/dask_cudf/dask_cudf/io/tests/test_parquet.py:1]: copyright

File: python/dask_cudf/dask_cudf/io/opt_parquet.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            column_names,
+            storage_name_mapping,
+            column_index_names,
+        ) = _parse_pandas_metadata(pandas_metadata)
+        categories = []
+        for col in pandas_metadata["columns"]:
+            if (col["pandas_type"] == "categorical") and (
+                col["name"] not in categories
+            ):
+                categories.append(col["name"])
```
[the remark this answers is not in the document]
Comment (rjzamora) [reply python/dask_cudf/dask_cudf/io/opt_parquet.py:87]: Not sure if it is worth using list comprehension here. We need to check if the next name is already in the new list before appending, and the list is not expected to be long.  If we avoid the loop-append approach, we would need to remove duplicates after the fact.

Let me know if I am just being slow here ( very likely :) )

[the remark this answers is not in the document]
Comment (kkraus14) [reply python/dask_cudf/dask_cudf/io/opt_parquet.py:93]: Ah I missed the reference in the loop to what we're appending to, my apologies. Agreed this looks good!

**Pull request merged**